crates/sui-data-ingestion-core/src/lib.rs

```diff
@@ -23,19 +24,19 @@ pub use worker_pool::WorkerPool;
 
 #[async_trait]
 pub trait Worker: Send + Sync {
-    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<()>;
-    /// Optional method. Allows controlling when workflow progress is updated in the progress store.
-    /// For instance, some pipelines may benefit from aggregating checkpoints, thus skipping
-    /// the saving of updates for intermediate checkpoints.
-    /// The default implementation is to update the progress store for every processed checkpoint.
-    async fn save_progress(
-        &self,
-        sequence_number: CheckpointSequenceNumber,
-    ) -> Option<CheckpointSequenceNumber> {
-        Some(sequence_number)
-    }
+    type Result: Send + Sync;
+    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<Self::Result>;
 
     fn preprocess_hook(&self, _: &CheckpointData) -> Result<()> {
         Ok(())
     }
 }
+
+#[async_trait]
+pub trait Reducer<R: Send + Sync>: Send {
+    async fn commit(&self, batch: Vec<R>) -> Result<CheckpointSequenceNumber>;
+    fn include_in_batch(&self, batch: &[R], element: &R) -> bool;
+    fn close_pending_batch(&self, _batch: &[R]) -> bool {
```

Comment:
Instead of the two interfaces, which I think is a bit difficult to reason about, we should have one interface called
`should_close_batch(&self, batch: &[R], next_item: Option<&R>) -> bool`
It looks at both the current batch state, as well as a potential next item that would be included in the batch if we don't close it, and decide if we should close the batch first.

Reply:
that's an option. Although I'm more leaning towards commit/partial_commit described above instead with a default implementation of partial_commit that just calls commit

Reply:
There won't be a `partial_commit` though. There will only be `commit` which always commit.
In the reducer however, each time when it needs to make a decision to commit, it calls `should_close_batch`.

Reply:
that's a proposed interface, not what currently is available in the PR.
So two methods commit/partial_commit. Nothing else. Second one is optional to implement with a default behavior of just aggregating greedy as much messages in batch as possible and committing them all together. 
Let me maybe update the PR to illustrate it better